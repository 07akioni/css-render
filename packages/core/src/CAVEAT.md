# CAVEAT
Some of internal functions are named using unreadable abbreviation, because it can slightly reduce the bundle size even after minify.
