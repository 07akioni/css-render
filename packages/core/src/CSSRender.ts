import {
  CNode
} from './types'
import {
  h,
  createCNode
} from './h'
import {
  render
} from './render'
import {
  mount
} from './mount'

export interface CSSRenderInstance {
  context: {
    [key: string]: any
  }
  id: string
  h: createCNode
  render: (node: CNode) => string
  mount: (nodes: CNode[] | CNode, id: string | number) => HTMLStyleElement
  use: (plugin: CSSRenderPlugin, ...args: any[]) => void
  config: CSSRenderConfig
}

export interface CSSRenderPlugin {
  install: (instance: CSSRenderInstance, ...args: any[]) => void
  [key: string]: any
}

interface CSSRenderConfig {
  preserveEmptyBlock: boolean
}

export function CSSRender (config: CSSRenderConfig = {
  preserveEmptyBlock: false
}): CSSRenderInstance {
  const cssr: CSSRenderInstance = {
    h,
    render: (node: CNode) => render(node, cssr),
    mount: (nodes: CNode[] | CNode, id: string | number) => mount(nodes, id, cssr),
    use: (plugin: CSSRenderPlugin, ...args: any[]) => plugin.install(cssr, ...args),
    id: Math.random().toString(36).slice(2, 10),
    context: {},
    config
  }
  return cssr
}
