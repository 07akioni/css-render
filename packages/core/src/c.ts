import {
  CNode,
  CProperties,
  CContext,
  CssRenderInstance,
  createCNodeForCssRenderInstance,
  baseCreateCNodeForCssRenderInstance,
  CRenderProps,
  CSelector,
  CNodeChildren,
  UnmountOption,
  MountOption
} from './types'
import { render } from './render'
import { mount, unmount } from './mount'

function wrappedRender <T extends CRenderProps> (this: CNode, props?: T): string {
  return render(this, this.instance, props)
}

const inNode = typeof window === 'undefined'

function wrappedMount (
  this: CNode,
  options: MountOption = {}
): HTMLStyleElement {
  /* istanbul ignore next */
  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
  if (inNode) return {} as HTMLStyleElement
  const { target, id } = options
  const { props, count = false } = options
  const targetElement = mount(
    this.instance,
    this,
    id ?? target,
    props,
    count
  )
  return targetElement as HTMLStyleElement
}

function wrappedUnmount (
  this: CNode,
  options: UnmountOption = {}
): void {
  /* istanbul ignore next */
  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
  if (inNode) return
  const {
    id,
    target,
    delay = 0,
    count = false
  } = options
  if (delay === 0) unmount(this.instance, this, id ?? target, count)
  else {
    setTimeout(() => unmount(this.instance, this, id ?? target, count), delay)
  }
}

const createCNode: baseCreateCNodeForCssRenderInstance = function (
  instance: CssRenderInstance,
  $: CSelector,
  props: CProperties,
  children: CNodeChildren
): CNode {
  return {
    instance,
    $,
    props,
    children,
    els: [],
    render: wrappedRender,
    mount: wrappedMount,
    unmount: wrappedUnmount
  }
}

export const c: createCNodeForCssRenderInstance = function (
  instance: CssRenderInstance,
  $: any,
  props: any,
  children: any
): CNode {
  if (Array.isArray($)) {
    return createCNode(instance, { $: null }, null, $)
  } if (Array.isArray(props)) {
    return createCNode(instance, $, null, props)
  } else if (Array.isArray(children)) {
    return createCNode(instance, $, props, children)
  } else {
    return createCNode(instance, $, props, null)
  }
} as createCNodeForCssRenderInstance

export {
  CNode,
  CProperties,
  CContext
}
