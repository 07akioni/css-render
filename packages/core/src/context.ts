import {
  CContext
} from './types'

export const context: CContext = {}
