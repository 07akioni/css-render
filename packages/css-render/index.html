<!DOCTYPE html>
<html>
  <head>
    <title>CssRender Test</title>
  </head>
  <body>
    <div class="a">123<div class="b">123</div></div>
    <script type="module" src="/playground/index.ts"></script>
  </body>
</html>