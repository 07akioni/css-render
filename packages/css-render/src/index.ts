/* istanbul ignore file */
import { CssRender } from './CssRender'

export * from './types'
export { default as hash } from './hash'
export { CssRender }
export default CssRender
