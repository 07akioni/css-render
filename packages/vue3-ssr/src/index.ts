import { inject, InjectionKey, App } from 'vue'

interface CssrSsrContext {
  styles: string[]
  ids: Set<string>
}

const ssrContextKey: InjectionKey<CssrSsrContext> = Symbol(
  '@css-render/vue3-ssr'
)

function createStyleString (id: string, style: string): string {
  return `<style cssr-id="${id}">\n${style}\n</style>`
}

function ssrAdapter (id: string, style: string): void {
  const ssrContext = inject(ssrContextKey, null)
  if (ssrContext === null) {
    console.error('[css-render/vue3-ssr]: no ssr context found.')
    return
  }
  const { styles, ids } = ssrContext
  // we need to impl other options to make it behaves the same as the client side
  if (ids.has(id)) return
  if (styles !== null) {
    ids.add(id)
    styles.push(createStyleString(id, style))
  }
}

export function useSsrAdapter ():
| {
  adapter: typeof ssrAdapter
  context: CssrSsrContext
}
| undefined {
  const context = inject(ssrContextKey, null)
  if (context === null) return undefined
  return {
    adapter: ssrAdapter,
    context
  }
}

interface SsrHandle {
  collect: () => string
}

export function setup (app: App): SsrHandle {
  const styles: string[] = []
  const ssrContext: CssrSsrContext = {
    styles,
    ids: new Set<string>()
  }
  app.provide(ssrContextKey, ssrContext)
  return {
    collect () {
      const res = styles.join('\n')
      styles.length = 0
      return res
    }
  }
}
