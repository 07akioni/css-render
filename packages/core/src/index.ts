import {
  CSelector,
  CNode,
  CProperties
} from './types'
import { context } from './context'
import { render } from './render'

interface createCNode {
  (path: string | CSelector): CNode
  (path: string | CSelector, children: CNode[]): CNode
  (path: string | CSelector, properties: CProperties): CNode
  (path: string | CSelector, properties: CProperties, children: CNode[]): CNode
}

export const h: createCNode = function (
  path: any,
  properties: any,
  children: any
) {
  if (properties === undefined) {
    return {
      path,
      properties: null,
      children: null
    }
  } else if (Array.isArray(properties)) {
    return {
      path,
      properties: null,
      children: properties
    }
  } else {
    if (Array.isArray(children)) {
      return {
        path,
        properties,
        children
      }
    } else {
      return {
        path,
        properties,
        children: null
      }
    }
  }
} as createCNode

export { context }
export { render }
