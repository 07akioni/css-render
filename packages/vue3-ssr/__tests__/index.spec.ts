import { CssRender } from 'css-render'
import { h, createSSRApp, defineComponent } from 'vue'
import { renderToString } from '@vue/server-renderer'
import { useSsrAdapter, setup } from '../src/index'

const { c } = CssRender()

describe('ssr', () => {
  describe('render to string', () => {
    const Child = defineComponent({
      setup () {
        c('div', {
          color: 'red'
        }).mount({
          id: 'mount-id',
          ssr: useSsrAdapter()
        })
      },
      render () {
        return h('div', null, 'Child')
      }
    })
    const App = defineComponent({
      render () {
        return h(Child)
      }
    })
    const app = createSSRApp(App)
    const { collect } = setup(app)
    it('should work', (done) => {
      renderToString(app).then((v) => {
        expect(collect() + v).toMatchSnapshot()
        done()
      })
    })
  })
  describe('useSsrAdapter', () => {
    const Child = defineComponent({
      setup () {
        c('div', {
          color: 'red'
        }).mount({
          id: 'mount-id',
          ssr: useSsrAdapter()
        })
      },
      render () {
        return h('div', null, 'Child')
      }
    })
    const App = defineComponent({
      render () {
        return h(Child)
      }
    })
    const app = createSSRApp(App)
    const { collect } = setup(app)
    it('should work', (done) => {
      renderToString(app).then((v) => {
        expect(collect() + v).toMatchSnapshot()
        done()
      })
    })
  })
  // uncomment after vue fixes the stream ssr bug
  // describe('render to stream', () => {
  //   it('should work', (done) => {
  //     const app = createSSRApp({
  //       render () {
  //         return [
  //           h(Foo),
  //           h(Suspense, null, {
  //             default: () => h(Bar),
  //             ssFallback: () => 'suspense'
  //           }),
  //           h(Foo)
  //         ]
  //       }
  //     })
  //     const Foo = {
  //       setup () {
  //         c('div', {
  //           color: 'foo'
  //         }).mount({
  //           id: 'foo',
  //           ssr: ssrAdapter
  //         })
  //       },
  //       render: () => h('div', 'foo')
  //     }
  //     const Bar = {
  //       async setup () {
  //         c('div', {
  //           color: 'bar'
  //         }).mount({
  //           id: 'bar',
  //           ssr: ssrAdapter
  //         })
  //         await new Promise((resolve) => setTimeout(resolve, 1000))
  //       },
  //       render: () => {
  //         return [h('div', 'bar'), h(Foo)]
  //       }
  //     }
  //     const { collect } = setup(app)
  //     const rs = renderToStream(app)
  //     let html = ''
  //     rs.on('data', (chunk) => {
  //       html += `${collect()}\n${chunk.toString() as string}`
  //     })
  //     rs.on('end', () => {
  //       expect(html).toMatchSnapshot()
  //       done()
  //     })
  //   })
  // })
})
