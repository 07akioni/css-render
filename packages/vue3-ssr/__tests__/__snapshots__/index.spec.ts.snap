// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr render to string should work 1`] = `
"<style cssr-id=\\"mount-id\\">
div {
  color: red;
}
</style><div>Child</div>"
`;

exports[`ssr useSsrAdapter should work 1`] = `
"<style cssr-id=\\"mount-id\\">
div {
  color: red;
}
</style><div>Child</div>"
`;
